gc.c

```diff
@@ -2466,6 +2408,47 @@ newobj_fill(VALUE obj, VALUE v1, VALUE v2, VALUE v3)
     return obj;
 }
 
+#if USE_RVARGC
+static inline VALUE
+heap_get_freeobj(rb_objspace_t *objspace, rb_size_pool_t *size_pool, rb_heap_t *heap)
+{
+    RVALUE *p = size_pool->freelist;
+
+    if (UNLIKELY(p == NULL)) {
+        struct heap_page *page = heap_next_freepage(objspace, size_pool, heap);
+        size_pool->using_page = page;
+
+        asan_unpoison_memory_region(&page->freelist, sizeof(RVALUE*), false);
+        p = page->freelist;
+        page->freelist = NULL;
+        asan_poison_memory_region(&page->freelist, sizeof(RVALUE*));
+        page->free_slots = 0;
+    }
+
+    asan_unpoison_object((VALUE)p, true);
+    size_pool->freelist = p->as.free.next;
+
+    return (VALUE)p;
+}
+
+static inline rb_size_pool_t *
+size_pool_for_size(rb_objspace_t *objspace, size_t size)
+{
+    size_t slot_count = CEILDIV(size, sizeof(RVALUE));
+
+    /* size_pool_idx is ceil(log2(slot_count)) */
+    size_t size_pool_idx = 64 - nlz_int64(slot_count - 1);
+    GC_ASSERT(size_pool_idx > 0);
+    GC_ASSERT(size_pool_idx < SIZE_POOL_COUNT);
```

Comment:
As this `size` comes from the caller, so it should be checked always, not only when `RGENGC_CHECK_MODE > 0`.
I'm not sure if it should `abort`, raise an exception or return `NULL`.

Reply:
I changed it to abort with `rb_bug` when the size is 0. Since it is a case that should never happen, I think aborting is the best option.

Reply:
That condition should be always false as the only caller checks it.
I meant `GC_ASSERT`s.
The former assertion is also always ok, from the range of `nlz_int64`.
My concern is the latter assertion.
`size_pool_idx` should be `SIZE_POOL_COUNT`(== 4) when `slot_count` is 16, that means `size > 8 * sizeof(RVALUE)`.
Since the `size` comes from callers of `rb_newobj` as `alloc_size` at runtime, it should not be a subject of "assertion" but should be checked at runtime too, regardless `RGENGC_CHECK_MODE`.

Reply:
Ah sorry, I think I misunderstood your original comment. Just to confirm, are you suggesting something like this?

```diff
    /* size_pool_idx is ceil(log2(slot_count)) */
    size_t size_pool_idx = 64 - nlz_int64(slot_count - 1);
+   if (size_pool_idx >= SIZE_POOL_COUNT) {
+       rb_bug("size_pool_for_size: allocation size too large");
+   }
```

Reply:
Yes, that is.
And “I'm not sure if it should `abort`, raise an exception or return `NULL`.”
Probably, the last option (`NULL`) will not be good, though.

Reply:
I'm thinking that this may be applicable to typed data too.
If it will be introduced, `alloc_size` can't be estimated.

Reply:
> I'm thinking that this may be applicable to typed data too.
> If it will be introduced, alloc_size can't be estimated.

I'm not sure I understand, could you elaborate further? When creating the typed data object using `TypedData_Make_Struct` we know the type of the struct required to be allocated, so wouldn't we know the exact size needed to be allocated?

Reply:
To know the exact size that can be allocated with this feature, `USE_RVARGC` and `SIZE_POOL_COUNT` are needed.
This is the reason I withdrew the previous comment.

Reply:
We plan on introducing API for callers to check whether a particular size is allocatable through VWA or not. If it's not, the caller is expected to fall back to some other way of allocating (e.g. using malloc).